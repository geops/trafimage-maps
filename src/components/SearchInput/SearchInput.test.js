import React from 'react';
import { render, fireEvent } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import userEvent from '@testing-library/user-event';
import SearchInput from './SearchInput';

describe('SearchInput', () => {
  describe('when no properties are set', () => {
    let spy = null;

    beforeEach(() => {
      window.console.error = jest.fn().mockImplementation(() => {});
      spy = jest.spyOn(window.console, 'error');
    });

    afterEach(() => {
      spy.mockRestore();
      window.console.error.mockRestore();
    });

    test('displays 1 error for required property ', () => {
      render(<SearchInput />);
      expect(spy).toHaveBeenCalledTimes(0);
    });

    test('matches snapshot', () => {
      const component = render(<SearchInput />);
      expect(component.container.innerHTML).toMatchSnapshot();
    });

    test('calls default onBlurInput, onFocus, onKeyPress without errors.', () => {
      const wrapper = render(<SearchInput />);
      const input = wrapper.container.querySelector('input');
      input.focus();
      input.blur();
      fireEvent.keyDown(input);
      fireEvent.keyUp(input);
    });
  });

  describe('when properties are set', () => {
    test('matches snapshot ', () => {
      const { container } = render(
        <SearchInput
          value="bar"
          className="tm-foo"
          placeholder="gux"
          onBlur={() => {}}
          onKeyPress={() => {}}
          onChange={() => {}}
        />,
      );
      expect(container.innerHTML).toMatchSnapshot();
      container.querySelector('input').focus();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test('matches snapshot when className is undefined ', () => {
      const { container } = render(<SearchInput className={undefined} />);
      expect(container.innerHTML).toMatchSnapshot();
    });

    test('calls onBlurInput, onFocus, onKeyPress properties.', () => {
      const fn = jest.fn();
      const { container } = render(
        <SearchInput onBlurInput={fn} onFocus={fn} onKeyPress={fn} />,
      );
      const input = container.querySelector('input');
      input.focus();
      input.blur();
      fireEvent.keyDown(input);
      fireEvent.keyUp(input);
      expect(fn).toHaveBeenCalledTimes(3);
    });

    // TODO to fix this test should work as it is
    test.skip('searchs on input change event.', () => {
      userEvent.setup();
      const fn = jest.fn();
      const { container } = render(<SearchInput onChange={fn} />);
      const input = container.querySelectorAll('input')[0];
      input.focus();
      userEvent.keyboard('test');
      fireEvent.keyUp(input, { target: { value: 'foo' } });
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test('searchs with an empty string on click on cross button.', () => {
      const fn = jest.fn();
      const { container } = render(<SearchInput onChange={fn} />);
      const input = container.querySelectorAll('input')[0];
      input.focus();
      fireEvent.click(container.querySelector('.tm-button'));
      expect(fn).toHaveBeenCalledTimes(1);
      expect(input).toBe(document.activeElement);
    });

    test('trigger onClickSearchButton callback.', () => {
      const fn = jest.fn();
      const { container } = render(<SearchInput onClickSearchButton={fn} />);
      const input = container.querySelectorAll('input')[0];
      input.focus();
      fireEvent.click(container.querySelectorAll('.tm-bt-search')[0]);
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test('trigger onClearClick callback.', () => {
      const fn = jest.fn();
      const { container } = render(<SearchInput onClearClick={fn} />);
      fireEvent.click(container.querySelectorAll('.tm-button')[0]);
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test('searchs on enter key up.', () => {
      userEvent.setup();
      const fn = jest.fn();
      const { container } = render(<SearchInput onChange={fn} />);
      const input = container.querySelectorAll('input')[0];
      input.focus();
      fireEvent.keyUp(input, { keyCode: 13, which: 13 });
      expect(fn).toHaveBeenCalledTimes(1);
    });

    test('focuses the input on state change.', () => {
      const { container } = render(<SearchInput />);
      const input = container.querySelectorAll('input')[0];
      input.focus();
      expect(input).toBe(document.activeElement);
    });
  });
});
