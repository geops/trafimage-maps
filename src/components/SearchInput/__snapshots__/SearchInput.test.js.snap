// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput when no properties are set matches snapshot 1`] = `
<div class="tm-search-input">
  <input type="text"
         tabindex="0"
         placeholder
         title
         value
  >
  <div role="button"
       style="display: none;"
       tabindex="0"
       class="tm-button tm-bt-clear"
       title
       aria-label
  >
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 24 24"
         focusable="false"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill="none"
            d="M0 0h24v24H0z"
      >
      </path>
      <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
      </path>
    </svg>
  </div>
  <div role="button"
       tabindex="0"
       class="tm-button tm-bt-search"
       title
       aria-label
  >
    search
  </div>
</div>
`;

exports[`SearchInput when properties are set matches snapshot  1`] = `
<div class="tm-foo">
  <input type="text"
         tabindex="0"
         placeholder="gux"
         title
         value="bar"
  >
  <div role="button"
       tabindex="0"
       class="tm-button tm-bt-clear"
       title
       aria-label
  >
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 24 24"
         focusable="false"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill="none"
            d="M0 0h24v24H0z"
      >
      </path>
      <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
      </path>
    </svg>
  </div>
  <div role="button"
       tabindex="0"
       class="tm-button tm-bt-search"
       title
       aria-label
  >
    search
  </div>
</div>
`;

exports[`SearchInput when properties are set matches snapshot  2`] = `
<div class="tm-foo">
  <input type="text"
         tabindex="0"
         placeholder="gux"
         title
         value="bar"
  >
  <div role="button"
       tabindex="0"
       class="tm-button tm-bt-clear"
       title
       aria-label
  >
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 24 24"
         focusable="false"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill="none"
            d="M0 0h24v24H0z"
      >
      </path>
      <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
      </path>
    </svg>
  </div>
  <div role="button"
       tabindex="0"
       class="tm-button tm-bt-search"
       title
       aria-label
  >
    search
  </div>
</div>
`;

exports[`SearchInput when properties are set matches snapshot when className is undefined  1`] = `
<div class="tm-search-input">
  <input type="text"
         tabindex="0"
         placeholder
         title
         value
  >
  <div role="button"
       style="display: none;"
       tabindex="0"
       class="tm-button tm-bt-clear"
       title
       aria-label
  >
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 24 24"
         focusable="false"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill="none"
            d="M0 0h24v24H0z"
      >
      </path>
      <path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z">
      </path>
    </svg>
  </div>
  <div role="button"
       tabindex="0"
       class="tm-button tm-bt-search"
       title
       aria-label
  >
    search
  </div>
</div>
`;
